components/eamxx/src/physics/gw/impl/gw_gw_ediff_impl.hpp

```diff
@@ -14,26 +17,107 @@ namespace gw {
 template<typename S, typename D>
 KOKKOS_FUNCTION
 void Functions<S,D>::gw_ediff(
-// Inputs
-const Int& ncol,
-const Int& pver,
-const Int& kbot,
-const Int& ktop,
-const uview_1d<const Int>& tend_level,
-const uview_1d<const Spack>& gwut,
-const uview_1d<const Spack>& ubm,
-const uview_1d<const Spack>& nm,
-const uview_1d<const Spack>& rho,
-const Spack& dt,
-const Spack& gravit,
-const uview_1d<const Spack>& pmid,
-const uview_1d<const Spack>& rdpm,
-const uview_1d<const Spack>& c,
-// Outputs
-const uview_1d<Spack>& egwdffi)
+  // Inputs
+  const MemberType& team,
+  const Workspace& workspace,
+  const Int& pver,
+  const Int& pgwv,
+  const Int& kbot,
+  const Int& ktop,
+  const Int& tend_level,
+  const Real& dt,
+  const uview_2d<const Real>& gwut,
+  const uview_1d<const Real>& ubm,
+  const uview_1d<const Real>& nm,
+  const uview_1d<const Real>& rho,
+  const uview_1d<const Real>& pmid,
+  const uview_1d<const Real>& rdpm,
+  const uview_1d<const Real>& c,
+  // Outputs
+  const uview_1d<Real>& egwdffi,
+  const uview_1d<Real>& decomp_ca,
+  const uview_1d<Real>& decomp_cc,
+  const uview_1d<Real>& decomp_dnom,
+  const uview_1d<Real>& decomp_ze)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  //-----------------------------Local Workspace------------------------------
+
+  // 0._r8 (array to pass to vd_lu_decomp).
+  Real zero = 0;
+  // Effective gw diffusivity at midpoints.
+  uview_1d<Real> egwdffm, tmpi2;
+  workspace.template take_many_contiguous_unsafe<2>(
+    {"egwdffm", "tmpi2"},
+    {&egwdffm, &tmpi2});
+
+  // Inverse Prandtl number.
+  static constexpr Real prndl=0.25;
+  // Density scale height.
+  static constexpr Real half=0.5;
+
+  //--------------------------------------------------------------------------
+
+  Kokkos::parallel_for(
+    Kokkos::TeamVectorRange(team, pver+1), [&] (const int k) {
+      egwdffi(k) = 0;
+      egwdffm(k) = 0;
+      tmpi2(k)   = 0;
+  });
+
+  team.team_barrier();
+
+  // Calculate effective diffusivity at midpoints.
+  const int num_pgwv = 2*pgwv + 1;
+  Kokkos::parallel_for(
+    Kokkos::TeamVectorRange(team, ktop+1, kbot+1), [&] (const int k) {
+      for (int l = 0; l < num_pgwv; ++l) {
+        egwdffm(k) += prndl * half * gwut(k,l) * (c(l)-ubm(k)) / bfb_square(nm(k));
+      }
+    });
```

Comment:
We should probably have an overload for `ekat::ReductionUtils::parallel_reduce()` for when you are in a `TeamThreadRange` which will serialize when ekatBFB=true.

Reply:
@tcclevenger , by "should have" do you mean you think we already have one or that we should create one?